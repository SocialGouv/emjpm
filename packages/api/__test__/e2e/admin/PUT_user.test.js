//

const request = require("supertest");
const nodemailerMock = require("nodemailer-mock");
jest.setMock("nodemailer", nodemailerMock);

// Fake env
process.env.SMTP_FROM = "ne-pas-repondre@emjpm.gouv.fr";
process.env.APP_URL = "https://emjpm.gouv.fr";

const { knex } = global;
jest.setMock("@emjpm/api/db/knex", knex);

const server = require("@emjpm/api/app");

const { getTokenByUserType } = require("../utils");

//

beforeAll(async () => {
  await knex.migrate.latest();
  await knex.seed.run();
});

beforeEach(() => {
  nodemailerMock.mock.reset();
});

afterEach(async () => {
  await knex("users")
    .where({ username: "jeremy" })
    .update({ active: true });
});

test("admin should be able to change jeremy (user 1) profile", async () => {
  {
    const user = await knex("users")
      .where({ id: 1 })
      .first();
    expect(user.active).toBeTruthy();
  }

  const token = await getTokenByUserType("admin");
  const response = await request(server)
    .put("/api/v1/admin/user/1")
    .set("Authorization", "Bearer " + token)
    .send({ active: false });

  expect(response.body).toMatchSnapshot({
    created_at: expect.any(String),
    password: expect.any(String)
  });

  expect(response.status).toBe(200);

  expect(nodemailerMock.mock.sentMail().length).toBe(1);
  expect(nodemailerMock.mock.sentMail()).toMatchSnapshot();

  {
    const user = await knex("users")
      .where({ id: 1 })
      .first();
    expect(user.active).toBeFalsy;
  }
});

test("fail because only 'active' key can be patched", async () => {
  const token = await getTokenByUserType("admin");
  const response = await request(server)
    .put("/api/v1/admin/user/1")
    .set("Authorization", "Bearer " + token)
    .send({ foo: "bar" });

  expect(response.body).toMatchInlineSnapshot(
    { stack: expect.any(String) },
    `
            Object {
              "message": "Not 'active' field to patch.",
              "name": "UnprocessableEntityError",
              "stack": Any<String>,
              "status": 422,
            }
      `
  );

  expect(response.status).toBe(422);

  expect(nodemailerMock.mock.sentMail().length).toBe(0);
});

test("fail because user 1024 is not found ", async () => {
  const token = await getTokenByUserType("admin");
  const response = await request(server)
    .put("/api/v1/admin/user/1024")
    .set("Authorization", "Bearer " + token)
    .send({ active: true });

  expect(response.body).toMatchInlineSnapshot(
    { stack: expect.any(String) },
    `
    Object {
      "message": "User \\"1024\\" not Found",
      "name": "NotFoundError",
      "stack": Any<String>,
      "status": 404,
    }
  `
  );

  expect(response.status).toBe(404);

  expect(nodemailerMock.mock.sentMail().length).toBe(0);
});
