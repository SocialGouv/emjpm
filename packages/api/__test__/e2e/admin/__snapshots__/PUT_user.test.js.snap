// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`admin should be able to change jeremy (user 1) profile 1`] = `
Object {
  "active": false,
  "cabinet": null,
  "created_at": Any<String>,
  "email": "ud@ud.com",
  "id": 1,
  "last_login": null,
  "nom": null,
  "password": Any<String>,
  "prenom": "Adrien",
  "reset_password_expires": null,
  "reset_password_token": null,
  "type": "individuel",
  "username": "jeremy",
}
`;

exports[`admin should be able to change jeremy (user 1) profile 2`] = `
Array [
  Object {
    "from": "ne-pas-repondre@emjpm.gouv.fr",
    "html": "
Bonjour,<br>
<br>
eMJPM vient de valider votre compte.
<br>
Vous pouvez l'utiliser.

Veuillez cliquer sur le lien suivant: <a href=https://emjpm.gouv.fr>Se connecter</a>
<br><br><br>
Bien à vous.
<br><br><br>
eMJPM Team
",
    "subject": "Validation du compte",
    "text": "
Bonjour,

eMJPM vient de valider votre compte.

Vous pouvez l'utiliser.

Veuillez cliquer sur le lien suivant: https://emjpm.gouv.fr

Bien à vous.

eMJPM Team
",
    "to": "ud@ud.com",
  },
]
`;
